Merge feature-emac branch into master

kjbracey opened the pull request:
### Description

Bring in the new `EMAC` network driver subsystem from the `feature-emac` branch. Primary PRs for work on that branch were #5558 and #5750.

Aside from EMAC-related work, also incorporates Non-blocking DNS (#6751) and default network interface support (#6855).

Full list of PRs on that branch: https://github.com/ARMmbed/mbed-os/pulls?q=is%3Apr+base%3Afeature-emac+is%3Aclosed

When this is merged, support for existing native lwIP and old-style emac drivers will be lost - only updated drivers will remain supported. Platforms losing existing drivers are:

* ARM_CM3DS_MPS2
* LPCTarget (LPC1768 family)
<strike>* RZ_A1XX</strike>

Some test reconfiguration will be required to get this all green, both to adjust setup and in case any platforms losing Ethernet support due to lack of driver update.

### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

kjbracey committed:
Networking update: general refactoring, unifying EMAC

Initial work by Bartek Szatkowski in https://github.com/ARMmbed/mbed-os/pull/4079,
reworked following review of https://github.com/ARMmbed/mbed-os/pull/5202 to
transform the entire system into C++, retaining the basic functionality.

Bartek's summary:

* Porting ethernet to EMAC
* Updating EMAC to enable multiple interfaces
* Untangling networking classes, making the abstractions a bit clearer to follow, etc
* General refactoring
* Removal of DEVICE_EMAC flag and introducing DEVICE_ETH and DEVICE_WIFI

Revisions since initial branch:

* Remove lwip depencies
* Correct doxygen warnings
* Remove emac_api.h, replace with C++ EMAC abstract class.
* Create OnboardNetworkInterface, and LWIP implementation.
* Mappings since #4079
     lwip-interface/nsapi_stack_lwip.c -> LWIPStack.cpp
     lwip-interface/ipstack_lwip.c -> LWIPInterface.cpp
     netsocket/mbed_ipstack.h -> OnboardNetworkStack.h
     hal/emac_api.h -> EMAC.h
* Reinstate use of EthInterface abstraction
* Correct and clarify HW address EMAC ops
* Restore MBED_MAC_ADDR implementation
* Integrate PPP support with LWIP::Interface.
* Convert K64F lwIP driver to K64F_EMAC.

To do:

* Convert emac_stack_mem.h to follow this pattern.
* Figure out DEVICE_ETH/EMAC
* Update all drivers to use EMAC

kjbracey committed:
Insert EMACInterface class

Rather than let "EthernetInterface" be the base EMAC NetworkInterface,
insert an "EMACInterface" class.

EthernetInterface then derives from EMACInterface and EthInterface.

A Wi-Fi driver can derive from EMACInterface and WiFiInterface - this
will be more logical than deriving from EthernetInterface and
WiFiInterface.

This does mean adding a couple of virtual inheritances to avoid
duplicate NetworkInterfaces:

                   NetworkInterface
                     /           \
            virtual /             \ virtual
                   /               \
             EMACInterface     WiFiInterface
                   \               /
                    \             /
                     \           /
                  MyCustomWiFiInterface

kjbracey committed:
Remove CellularInterface

This has been superceded by CellularBase. Name change occurred late
in review of https://github.com/ARMmbed/mbed-os/pull/4119 and
original unused CellularInterface was left behind.

kjbracey committed:
Nanostack EMAC implementation

Make Nanostack an OnboardNetworkInterface, implementing
add_ethernet_interface so it can use EMAC drivers.

Can now be used via EthernetInterface, and be the system's default
network stack.

Legacy support for NanostackEthernetInterface retained. Some
restructuring of mesh interface code to fit into the
OnboardNetworkStack:::Interface system.

kjbracey committed:
Add downcast methods to NetworkInterface

As we've introduced virtual inheritance to support EMACInterface, we can
no longer use C-style casts or static_cast to downcast from
NetworkInterface to more specific types. RTTI is disabled in the
toolchains, so dynamic_cast is unavailables.

Add virtual downcast methods to permit conversions to the 6 derived
classes. Probably only needed for EMACInterface, WiFiInterface and
EthInterface, but handles the set.

Commit:
Created memory manager class to netsocket and updated lwip to use it
instead of old memory interface.

Commit:
Updated K64F ethernet driver to use memory manager

Commit:
Updated STM ethernet driver to use emac and memory manager

Commit:
Updated greentea network interface configuration files

kjbracey committed:
EMAC: check link status callback is set

Nanostack doesn't set the link status callback. Make sure the two
example drivers don't crash if it isn't set.

Commit:
Added preferred alignment to emac and copy to/from to memory manager

kjbracey committed:
Extend EMAC multicast APIs

kjbracey committed:
K64F EMAC: Fix TX error path leaks

kjbracey committed:
Move Freescale EMAC driver out of lwIP

kjbracey committed:
Move STM EMAC driver out out lwIP

kjbracey committed:
Tell mbed build to ignore old lwIP+EMAC drivers

kjbracey committed:
Work around Nuvoton #define EMAC

kjbracey committed:
Adjust test configurations for EMAC

Make ETHERNET configuration the default if DEVICE_EMAC is present,
instead of if FEATURE_LWIP is present.

This limits it to targets which have been ported to the new EMAC API.

Add LWIP feature to JSON config, as in principle the targets shouldn't
be adding it themselves. Opens scope to having Nanostack-based tests.

Disable tests for the Realtek and Wifi drivers that aren't ported yet.

kjbracey committed:
Restrict client and socket example tests to K64F+K66F

I would like to restrict these to devices with "device_has": "EMAC", but
the framework doesn't currently permit that.  Revisit as more drivers
are EMAC-enabled or if the framework changes.

kjbracey committed:
Disable Nanostack border router test

Pending Nanostack EMAC work, disable the border router.

kjbracey committed:
Ignore old EMAC Wifi drivers

Suppress Odin W2 and Realtek Wi-fi drivers using .mbedignore

kjbracey committed:
Remove Ethernet from NUCLEO_F429ZI test

Pending EMAC driver update, these tests won't compile.

kjbracey committed:
Connectivity test: don't create interface object twice

Commit:
Enabled greentea and mbed-os-example-tls/client/sockets test for STM

kjbracey committed:
Add NetworkInterface::get_default_instance()

Provide an initial framework to make it easier to find a default network
interface.

Commit:
Emac greentea tests for feature-emac

Commit:
Corrected STM eth driver flagging, memory allocation and thread init

Commit:
Made feature-emac changes to emac greentea tests default code

Removed flagging of the new/old code.

kjbracey committed:
Stop drivers checking DEVICE_EMAC

Checking DEVICE_EMAC is problematic - particularly for the Odin W2 where
apps have been shutting this off to disable the Wi-fi interface.

Make drivers check a locally-relevant flag instead, pending new
thoughts on how to achieve application/test-relevant flagging for
XXX:get_default_instance() being provided by a system.

However that is achieved, drivers do require a flag set purely by the
target - they mustn't be tripped up by an add-on module providing itself
as the system's default EMAC.

kjbracey committed:
Add extra labels for Freescale & STM EMAC drivers

Just checking "does the chip have an EMAC" doesn't work - there are
targets using those chips which do not have an Ethernet connector and
don't provide the necessary surrounding infrastructure (eg DISCO_F429ZI,
not providing the board emac config call, and HEXIWEAR not providing PHY
info).

Make the targets that actually do want EMAC define their own local
Freescale_EMAC and STM_EMAC labels, and move the drivers into
the corresponding TARGET_ directories, removing the #ifdefs.

kjbracey committed:
Add EMAC driver README.md with porting guide

Commit:
Changed STM emac driver to loop RX frame reading

Commit:
Added non-blocking DNS functionality to network interface

- Added non-blocking DNS interface to network interface and
  network stack.
- Added caching of DNS replies.
- Added a network stack function to get DNS addresses from stack.
- Added call and call_in hooks to onboard network stack to
  allow calling functions from onboard stack context.
- Added support to call and call_in functions to LWIP and
  Nanostack.
- Disabled LWIP DNS translator with the exception of DNS
  address storage used in DNS address get.

mmahadevan108 committed:
K64F: Updated the SYSMPU SDK driver

This is required to setup the MPU for ENET bus master accesses

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
K64F, K66F: Update the SYSMPU implementation

Do not disable SYSMPU, instead add access for ENET bus master

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
Kinetis EMAC: Renamed K64F files to Kinetis

This driver is used by more SoC's than K64F

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

Commit:
Updated non-blocking connect changes to EMAC greentea test server

mmahadevan108 committed:
LPC546XX: Add ENET support

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

Commit:
Corrected defects
- Changed call_in/call methods of the stack to callback provided by the stack
- Specified what are limitations for operations that are made in callback
- Added pure virtual class DNS that defines DNS operations
- Added cancel operation and unique ID to DNS request used in cancel
- Added DNS configuration options to netsocket/mbed_lib.json for retries,
  response wait time and cache size
- Changed host name to use dynamic memory in DNS query list and cache,
  set maximum length for the name to 255 bytes.
- Added mutex to asynchronous DNS
- Reworked retries: there is now total retry count and a server specific count
- Ignores invalid incoming UDP socket messages (DNS header is not valid), and retries DNS query
- Reworked DNS module asynchronous operation functions
- Corrected other review issues (nothrow new, missing free, missing mutex unlock etc.)

mmahadevan108 committed:
Add memory configuration for LPC546XX

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

cyliangtw committed:
[M487/NUC472] Support new EMAC feature, built OK

cyliangtw committed:
[M487/NUC472] Support DHCP & TCP sample, also remove legacy emac

Commit:
Updated emac greentea tests
- Improved markdown
- Set tests to fail if CTP echo server is not present

Commit:
Corrected more defects

- Serialized the sending of multiple async DNS queries since limits on event
message box sizes
- Added timer function that supervises the total time used to make DNS query
and triggers socket timeouts
- Changed nsapi_error_t to new nsapi_value_or_error_t on interface headers
- Corrected wording of gethostbyname_async return values
- Clarified .json options
- Added a new data type for socket callback that can be used from interrupts
- Corrected variable limits to use INT32_MAX etc. defines
- Changed mallocs to new
- Optimized variable sizes on DNS_QUERY definition

Commit:
Copy edit README.md

Copy edit changes to README.md, mostly for active voice, consistent tense across docs and style.

cyliangtw committed:
[M487/NUC472] Fix tx frame length and remove GPL license

TomoYamanaka committed:
Implementation of unified EMAC driver for Renesas mbed boards

Implementation of unified EMAC driver for Renesas mbed boards

Based on the driver so far, Renesas implemented the emac driver for GR-PEACH and VK-RZ/A1H.
The mainly changes is below.
- Add the connection part with LWIP according to the unified emac specification.
- Add three new multicast functions(add, remove, set_all).

The Greentea test netsocket and emac test passed.

cmonr commented:
@kjbracey-arm Is the goal to bring this in for 5.9?

kjbracey commented:
@cmonr - yes, the aim is to get this in this week for 5.9

Commit:
Changed mutexes, delete and DNS call in callback set

- Changed mutexes to platform mutexes
- Removed not needed null check from delete
- Changed nsapi_dns_call_in_set() to use call_in_callback_cb_t and
added prototype to header

Commit:
EMAC adaption added, updated ODIN drivers to v2.5.0 RC1

Commit:
in ODIN emac initialization required before connection

Commit:
Corrected lwip netif flags to be set before power up

This enables that link up status callback can be called inside
power up function.

kjbracey commented:
/morph build

Commit:
Added memory manager set to add_ethernet_interface() of test stack

This allows that EMAC memory manager can be used to allocate EMAC driver
memory already after the add_ethernet_interface() call.

Commit:
Removed EMAC get default instance from EMAC tests

Test environment now uses the EMAC defined by add ethernet interface.

cmonr commented:
Making a note here that this will need a PR for its example as well as a handbook PR for documentation, if not already available.

kjbracey committed:
LPC546XX: Correct Ethernet MAC address write

Patch to LPC546XX SDK code - write the low Ethernet MAC address
register last, as that synchronises the update.

Without this change, the ENET_SetMacAddr call only seems to work prior
to MAC initialisation, causing problems for the new mbed OS EMAC system,
which expects it to be changable later.

Updated emac greentea tests #6851.

kjbracey committed:
LPC546XX: Correct Ethernet length calculations

Subtract 4 from the received packet length - the buffer contains the
CRC, which we shouldn't pass up.

Ensure we allocate receive buffers of a size corresponding to the
rounded-up size we tell the hardware - the hardware was overrunning the
allocation by a couple of bytes.

Commit:
WIFI_EMAC class renamed to OdinWiFiEMAC, Formatting
Revert "in ODIN emac initialization required before connection"

M-ichae-l committed:
add-rtl8195am-feature-emac (#6904)

rtl8195am feature emac implementation.

Commit:
WiFi EMAC class name reflected in WiFi drivers binaries

Commit:
Updated ublox odin greentea test configuration file connect statement

Statement now uses network interface wifiInterface() method to
get wifi interface instead of casting.

Commit:
recompiled driver against NetworkInterface changes on latest feature-emac

geky commented:
@kjbracey-arm, it looks like the mbed 2 compilation test in events is trying to compile the features/network folder without the tracing folder. Should this list be updated?
https://github.com/ARMmbed/mbed-os/blob/master/.travis.yml#L153

Can we just remove the entire features folder for the test?

cmonr commented:
@kjbracey-arm Also, take a look at the pr-head build. If you don't have access or need help navigating the Jenkins links, let us know.

The error(s):
```
[UBLOX_EVK_ODIN_W2_arm-none-eabi-gcc_wlan0] ./mbed-os/targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_MODULE_UBLOX_ODIN_W2/sdk/TOOLCHAIN_GCC_ARM/libublox-odin-w2-driver.a(OdinWiFiInterface.o):(.rodata._ZTC17OdinWiFiInterface4_13EMACInterface+0x110): undefined reference to virtual thunk to EMACInterface::get_mac_address()
```

kegilbert commented:
Besides the current merge conflicts/CI errors was there anything in particular you'd like me to review? A (very) quick skim from a code perspective looks good to me.

kjbracey commented:
GitHub display went weird - rebased feature-emac to clear it up. Shouldn't cause a problem, as no open PRs to it.

hasnainvirk commented on features/netsocket/mbed_lib.json:
```diff
@@ -1,6 +1,34 @@
-        "present": 1
+        "present": 1,
+        "default-stack": "LWIP",
+        "default-interface-type": {
```
This shouldn't be here.  Seemingly, its added to targets.json already.

hasnainvirk commented on targets/targets.json:
```diff
@@ -285,7 +295,10 @@
-        "bootloader_supported": true
+        "bootloader_supported": true,
+        "overrides": {
+            "network-default-interface-type": "ETHERNET"
```
This should be CELLULAR by default.

SeppoTakalo replied:
Why?
C027 does have ethernet, and ethernet does not require any configuration. Therefore it should be the default.

For cellular to work, you need to provide configuration.

kjbracey replied:
Cellular can work without configuration - we have a default APN database.

hasnainvirk commented on targets/targets.json:
```diff
@@ -2094,14 +2143,17 @@
-        "bootloader_supported": true
+        "bootloader_supported": true,
+        "overrides": {
+            "network-default-interface-type": "ETHERNET"
```
This should be WIFI by default

hasnainvirk commented on targets/targets.json:
```diff
@@ -2160,11 +2212,14 @@
-        "bootloader_supported": true
+        "bootloader_supported": true,
+        "overrides": {
+            "network-default-interface-type": "ETHERNET"
```
This should be CELLULAR by default.

hasnainvirk commented on features/netsocket/EMAC.h:
```diff
@@ -148,7 +148,7 @@ class EMAC {
- *  be used by the EthernetInterface class's default constructor to initialise the networking subsystem.
+ *  be used by the EMACInterface class's default constructor to initialise the networking subsystem.
  */
 //extern const emac_interface_ops_t mbed_emac_eth_ops_default;
```
Dead code ? Should be remove I guess.

hasnainvirk commented on features/netsocket/CellularBase.h:
```diff
@@ -101,6 +101,7 @@ class CellularBase: public NetworkInterface {
      */
     virtual const char *get_gateway() = 0;
 
+    virtual CellularBase *cellularBase() { return this; }
```
This will fail in Travis astyle job. One liners are not allowed.

hasnainvirk commented on features/netsocket/EMAC.h:
```diff
@@ -44,8 +44,8 @@ class EMAC {
      */
-    //typedef void (*emac_link_input_fn)(void *data, emac_stack_mem_chain_t *buf);
-    typedef mbed::Callback<void (emac_stack_mem_chain_t *buf)> emac_link_input_cb_t;
+    //typedef void (*emac_link_input_fn)(void *data, emac_mem_buf_t *buf);
```
Dead code could be removed.

hasnainvirk commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_MODULE_UBLOX_ODIN_W2/sdk/wifi_emac/.mbedignore:
```diff
@@ -0,0 +1,2 @@
+wifi_emac_api.cpp
```
Are these drivers going to reside in the tree for a while ? I mean if they are already replaced by new implementation, they could just go away.

hasnainvirk commented on features/netsocket/EMACMemoryManager.h:
```diff
@@ -36,6 +36,8 @@
  *
  */
 
+#include "nsapi.h"
```
This inclusion doesn't have any effect on the immediate header. It doesn't seem to belong here.

hasnainvirk commented on features/nanostack/mbed-mesh-api/mbed-mesh-api/NanostackEMACInterface.h:
```diff
@@ -0,0 +1,29 @@
+ * limitations under the License.
+ */
+
+#ifndef NANOSTACKETHERNETINTERFACE_H
```
Header inclusion check is wrong.  This should be NANOSTACKEMACINTERFCAE_H.
You already have NANOSTACKETHERNETINTERFACE_H in NanostackEthernetInterface.h

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/MeshInterfaceNanostack.cpp:
```diff
@@ -15,56 +15,100 @@
+
+    if (buflen >= 40 && arm_net_address_get(interface_id, ADDR_IPV6_GP, binary_ipv6) == 0) {
+        ip6tos(binary_ipv6, buf);
+        //tr_debug("IP address: %s", address);
```
commented out code could be removed.

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackEMACInterface.cpp:
```diff
@@ -0,0 +1,221 @@
+#include "nsdynmemLIB.h"
+#include "arm_hal_phy.h"
+
+#if 1
```
Please remove #if 1

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackEMACInterface.cpp:
```diff
@@ -0,0 +1,221 @@
+    virtual void set_mac_address(uint8_t *mac);
+
+    int8_t address_write(phy_address_type_e , uint8_t *);
+    int8_t tx(uint8_t *data_ptr, uint16_t data_len, uint8_t tx_handle,data_protocol_e data_flow);
```
the name could emac_phy_tx() in correspondence with emac_phy_rx()

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackEMACInterface.cpp:
```diff
@@ -0,0 +1,221 @@
+static int8_t emac_phy_address_write(phy_address_type_e address_type, uint8_t *address_ptr)
+{
+    return single_phy->address_write(address_type, address_ptr);
+
```
extra line

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackEMACInterface.cpp:
```diff
@@ -0,0 +1,221 @@
+    return -1;
+}
+
+static int8_t emac_phy_tx(uint8_t *data_ptr, uint16_t data_len, uint8_t tx_handle,data_protocol_e data_flow)
```
The name of this function could be tx(), and the one exposed as an API should be emac_py_tx(). 
data_ptr should be const.

kjbracey replied:
Naming scheme appears to be that all the static functions we're providing as Nanostack callback functions are `emac_phy_`. Not sure I see a problem with that - I think they're consistent at least.

We can't change the prototype, alas - they're very historical Nanostack driver attachment points, used by third party drivers. Maybe worth attempting, but would need a new Nanostack release, and modifying all drivers, including this glue later.

And the public API including the 'tx' is inherited from `NanostackPhy` which is also public API being implemented by third parties

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackEthernetInterface.cpp:
```diff
@@ -18,67 +18,104 @@
 }
 
-int NanostackEthernetInterface::connect()
+#if 0
```
Remove dead code.

hasnainvirk commented on features/nanostack/mbed-mesh-api/source/NanostackRfInterface.cpp:
```diff
@@ -0,0 +1,11 @@
+
+#include "NanostackRfInterface.h"
+
+/*NanostackRfInterface::NanostackRfInterface(NanostackRfPhy &phy) :
```
Dead code could be removed.

hasnainvirk commented on features/nanostack/nanostack-interface/NanostackInterface.h:
```diff
@@ -19,225 +19,11 @@
 #define NANOSTACK_INTERFACE_H_
```
Why this header 'NanostackInterface.h' exists. It is just including other headers, a part which can be moved to its appropriate user class.

kjbracey replied:
Backwards compatibility - various apps include it.

hasnainvirk reviewed (dismissed):
Just a few comments.

cmonr commented:
Fyi, pr-head is still not passing. If you need help going through it, let me know.

kjbracey commented:
/morph build

kjbracey commented:
@hasnainvirk - addressed or responded to everything, I think.

hasnainvirk reviewed (dismissed):
Looks good to me.

kjbracey commented:
/morph build

kjbracey committed:
Refill example lists - all current Ethernet boards

Remove check for LWIP - check target flagging for EMAC drivers.
Should still be DEVICE_EMAC, but tooling doesn't support that.

kjbracey commented:
/morph build

kjbracey commented:
Couple of nanostack border router test failures - done 1 update to border router app to fix K64F and disabled the test for STMF429 - that's not resolvable, as Nanostack and EMAC driver can't co-exist.

Nanostack border router app will be updated to use EMAC drivers once this is merged to master: 
https://github.com/ARMmbed/nanostack-border-router/pull/85

kjbracey commented:
/morph build

cmonr commented:
Restarting pr-head. License checkout issue.

cmonr commented:
@kjbracey-arm 
```
"/tmp/d2gAAH", line 124 (column 3): Warning: L6312W: Empty Execution region description for region RW_IRAM1
Error: L6218E: Undefined symbol OnboardNetworkStack::get_default_instance() (referred from BUILD/K64F/ARM/mbed-os/features/netsocket/EthernetInterface.o).
Finished: 0 information, 3 warning and 1 error messages.
```
This is with the K64F nanostack-border-router example

kjbracey commented:
Okay, just going to have to kill the border router test again. Root cause here is that you can't have DEVICE_EMAC enabled on an ARMCC build if you don't have a default network stack.

Border router app is not touching the network stack setting, so it's LWIP, but it's removing FEATURE_LWIP, so left with no default stack. We can't fix this up until after the PR is merged, when it can switch the default network stack to Nanostack.

kjbracey commented:
/morph build

cmonr commented:
@kjbracey-arm Good news and bad news.

The bad news is that this hit a test issue that is in the process of being resolved here: https://github.com/ARMmbed/mbed-os/pull/6993

The good news is that after a rebase once that PR is in, it looks like CI should pass!

0xc0170 commented:
/morph build
